tests/unit/lib/SGMatrix_unittest.cc

```diff
@@ -761,3 +761,45 @@ TEST(SGMatrixTest,iterator)
 	for (auto v: mat)
 		EXPECT_EQ(mat[index++], v);
 }
+
+TEST(SGMatrixTest,column_iterator)
+{
+	constexpr index_t size=5;
+	SGMatrix<float64_t> mat(size, size);
+	linalg::range_fill(mat, 1.0);
+
+	auto begin_col = mat.begin_column();
+	auto end_col = mat.end_column();
+
+	EXPECT_EQ(mat.num_cols, std::distance(begin_col, end_col));
+	EXPECT_EQ(mat.get_column(0),*begin_col);
+	++begin_col;
+	EXPECT_EQ(mat.get_column(1),*begin_col);
+	--begin_col;
+	EXPECT_EQ(mat.get_column(0),*begin_col++);
+	EXPECT_TRUE(begin_col != end_col);
+	++begin_col;
+	EXPECT_EQ(mat.get_column(2),*begin_col--);
+	--begin_col;
+	EXPECT_EQ(mat.get_column(0),*begin_col);
+	begin_col += 2;
+	EXPECT_EQ(mat.get_column(2),*begin_col);
+	begin_col -= 1;
+	EXPECT_EQ(mat.get_column(1),*begin_col);
+	EXPECT_EQ(mat.get_column(2),begin_col[1]);
+	
+	auto new_itr = begin_col + 2;
+	EXPECT_EQ(mat.get_column(3),*new_itr);
+
+	//range-based loop should work as well
+	auto index = 0;
+	for(const auto& i : mat.columns())
+		EXPECT_EQ(i,mat.get_column(index++));
+	
+	const SGMatrix<float64_t> const_mat(mat);
+
+	index = 0;
+	for(const auto& i : const_mat.columns())
+		EXPECT_EQ(i,const_mat.get_column(index++));
+}
```

Comment:
what about testing for modifying the value of the column vector?
```cpp
auto mat_copy = mat.clone();
for(const auto& vec : mat_copy.columns())
	linalg::add(vec, 1, vec); //can't remember signature but basically vec += 1
for(const auto& [result, original] : zip_iterator(mat_copy.columns(), mat.columns()))
{
        auto reference = linalg::add(original) // again can't remember signature but something like reference = original + 1 
	EXPECT_EQ(result, reference);
}
```

Reply:
It's this
```c++
1	//Modifying value of column vector
2	auto mat_copy = mat.clone();
3
4	for(const auto& vec : mat_copy.columns())
5		mat_copy.set_column(index,linalg::add(vec,vec));
6	
7	for(const auto& [result, original] : zip_iterator(mat_copy.columns(),mat.columns()))
8	{
9			auto reference = linalg::add(original,original)
10			EXPECT_EQ(result, reference);
11	}
```
However zip_iterator is showing at line 7. Here's the pastebin https://pastebin.com/DfjXGkuR
```
no instance of constructor "zip_iterator" matches the argument list -- argument types are: (shogun::SGMatrix<float64_t>::column_struct<shogun::SGMatrix<float64_t> &>, shogun::SGMatrix<float64_t>::column_struct<shogun::SGMatrix<float64_t> &>)
```